packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp

```diff
@@ -115,12 +119,81 @@ namespace MueLu {
       const RCP<const Vector> D = (!fixing ? Utilities::GetInverse(diag) : Utilities::GetInverse(diag, 1e-4, one));
       Ainv = MatrixFactory::Build(D);
     }
+    else if(method=="spai") {
+      Ainv = GetSparseInverse(A, A->getCrsGraph());
+    }
 
     GetOStream(Statistics1) << "Approximate inverse calculated by: " << method << "." << std::endl;
+    GetOStream(Statistics1) << "Ainv has " << Ainv->getGlobalNumRows() << "x" << Ainv->getGlobalNumCols() << " rows and columns." << std::endl;
 
     Set(currentLevel, "Ainv", Ainv);
   }
 
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>>
+  InverseApproximationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetSparseInverse(const RCP<Matrix>& Aorg, const RCP<const CrsGraph>& sparsityPattern) const {
+
+    // construct the inverse matrix with the given sparsity pattern
+    RCP<Matrix> Ainv = MatrixFactory::Build(sparsityPattern);
+    Ainv->resumeFill();
+
+    // gather missing rows from other procs to generate an overlapping map
+    RCP<Import> rowImport = ImportFactory::Build(sparsityPattern->getRowMap(), sparsityPattern->getColMap());
+    RCP<Matrix> A = MatrixFactory::Build(Aorg, *rowImport);
+
+    // loop over all rows of the inverse sparsity pattern (this can be done in parallel)
+    for(size_t k=0; k<sparsityPattern->getLocalNumRows(); k++) {
```

Comment:
Does this mean, that this loop can be further parallelized in the future step?

Reply:
Yes, using Kokkos would be an option.

Reply:
Let's keep this in mind for future work.